DOCS-2832: Move upload data to how-to (#3367)
viamrobotics/docs: docs/how-tos/upload-data.md

---
title: "How to Upload a Batch of Data"
linkTitle: "Upload a batch of data"
description: "Upload data to the Viam app from your local computer or mobile device using the data client API, Viam CLI, or Viam mobile app."
weight: 100
type: "docs"
tags: ["data management", "cloud", "sync"]
image: "/services/icons/data-capture.svg"
imageAlt: "Upload data to the Viam app"
images: ["/services/icons/data-capture.svg"]
aliases:
  - /data/upload/
  - /services/data/upload/
---

If you configured [data capture](/services/data/capture/) on your machine, Viam automatically uploads data to the cloud from the directory you specified, at the interval you specified.
However, if you want to upload a batch of data once from somewhere else, either from a different directory on your machine or from your personal computer or mobile device, you have several options:

> - Configure the path to your directory in the Viam app, wait for the files to sync, then delete the path.

Reviewer: ```suggestion
- Configure the path to your directory as an additional sync path for the machine's data management service, wait for the files to sync, then delete the path.
```

> However, if you want to upload a batch of data once from somewhere else, either from a different directory on your machine or from your personal computer or mobile device, you have several options:
> 
> - Configure the path to your directory in the Viam app, wait for the files to sync, then delete the path.
>   This option requires that you have `viam-server` installed on the machine.
> - Run a Python script to upload files from a folder using the data client API `file_upload_from_path` method.
>   You can do this on a computer that doesn't have `viam-server` installed on it.
> - Upload images from your mobile device using the Viam mobile app.
> 
> {{% alert title="In this page" color="tip" %}}
> 
> - [Sync a batch of data from another directory](#sync-a-batch-of-data-from-another-directory)
> - [Upload data with the Python SDK](#upload-data-with-python)
> - [Upload images with the Viam mobile app](#upload-images-with-the-viam-mobile-app)
> 
> {{% /alert %}}

Reviewer: I think this is duplicated and doesn't need to be. You can tweak it a bit but I think this keeps it more concise.

```suggestion
However, if you want to upload a batch of data once from somewhere else, either from a different directory on your machine or from your personal computer or mobile device, you have several options using the Viam app, the data client API, or the Viam mobile app.

{{% alert title="In this page" color="tip" %}}

- [Sync a batch of data from another directory](#sync-a-batch-of-data-from-another-directory) by configuring the path to the directory as an additional sync path in a machine's data management service. This requires the data to be on a machine running `viam-server`.
- [Upload data with the Python SDK](#upload-data-with-python) by running a Python script to upload files from a folder.
- [Upload images with the Viam mobile app](#upload-images-with-the-viam-mobile-app) from your mobile device.

{{% /alert %}}
```

Author: I thought the same thing but feared doing it in this long form way would be deemed to be inconsistent with the other pages :D Since you agree I'm happy to make this change

{{% /tablestep %}}
{{% tablestep %}}
**2. Configure sync from the additional folder**

In the **Additional paths**, enter the full path to the directory where the data you want to upload is stored, for example, `/Users/Artoo/my_cat_photos`.

Toggle **Syncing** to on (green) if it isn't already on.

{{<imgproc src="/services/data/data-sync-temp.png" resize="x1100" declaredimensions=true alt="Data service configured in the Viam app as described." >}}

Click **Save** in the top right corner of the page.

{{% /tablestep %}}
{{% tablestep %}}
**3. Confirm that your data uploaded**

Navigate to your [**DATA** page in the Viam app](https://app.viam.com/data/view) and confirm that your data appears there.
If you don't see your files yet, wait a few moments and refresh the page.

{{% /tablestep %}}
{{% tablestep %}}
**4. Remove the folder path**

Once the data has uploaded, navigate back to your data service config.
You can now delete the additional path you added.
You can also turn off **Syncing** unless you have other directories you'd like to continue to sync from.
{{% /tablestep %}}
{{< /table >}}

## Upload data with Python

> You can use the Viam Python SDK's data client API [`file_upload_from_path`](/appendix/apis/data-client/#fileuploadfrompath) method to upload one or more files from your computer to the Viam cloud.

Reviewer: Just shortening a bit
```suggestion
You can use the Python data client API [`file_upload_from_path`](/appendix/apis/data-client/#fileuploadfrompath) method to upload one or more files from your computer to the Viam cloud.
```

```sh {class="command-line" data-prompt="$"}
pip install viam-sdk
```

{{< /expand >}}

### Instructions

{{< table >}}
{{% tablestep link="/appendix/apis/data-client/" %}}
**1. Write connection code**

Create a Python script file in a directory of your choice and [add code to establish a connection](/appendix/apis/data-client/#establish-a-connection) from your computer to your [Viam app](https://app.viam.com) {{< glossary_tooltip term_id="location" text="location" >}} or individual {{< glossary_tooltip term_id="part" text="machine part" >}}.

{{% /tablestep %}}
{{% tablestep link="/appendix/apis/data-client/" %}}
**2. Add a `file_upload_from_path` API call**

Use the `file_upload_from_path` method to upload your data, depending on whether you are uploading one or multiple files:

{{< tabs >}}
{{% tab name="Upload a single file" %}}

To upload just one file, make a call to `file_upload_from_path` according to [the data client API documentation](/appendix/apis/data-client/#fileuploadfrompath).
The following example code could be placed inside the `main()` function, or inside a function called from `main()`:

```python {class="line-numbers linkable-line-numbers"}

> await data_client.file_upload_from_path(
>   # The ID of the machine part the file should be associated with
>   part_id="abcdefg-1234-abcd-5678-987654321xyzabc",
>   # Any tags you want to apply to this file
>   tags=["cat", "animals", "brown"],
>   # Path to the file
>   filepath="/Users/Artoo/my_cat_photos/brown-cat-on-a-couch.png"
> )

Reviewer: ```suggestion
import asyncio

from viam.rpc.dial import DialOptions, Credentials
from viam.app.viam_client import ViamClient

async def connect() -> ViamClient:
    dial_options = DialOptions(
      credentials=Credentials(
        type="api-key",
        # Replace "<API-KEY>" (including brackets) with your machine's API key
        payload='<API-KEY>',
      ),
      # Replace "<API-KEY-ID>" (including brackets) with your machine's
      # API key ID
      auth_entity='<API-KEY-ID>'
    )
    return await ViamClient.create_from_dial_options(dial_options)

async def main():
    # Make a ViamClient
    viam_client = await connect()
    # Instantiate a DataClient to run data client API methods on
    data_client = viam_client.data_client
    await data_client.file_upload_from_path(
      # The ID of the machine part the file should be associated with
      part_id="abcdefg-1234-abcd-5678-987654321xyzabc",
      # Any tags you want to apply to this file
      tags=["cat", "animals", "brown"],
      # Path to the file
      filepath="/Users/Artoo/my_cat_photos/brown-cat-on-a-couch.png"
    )

    viam_client.close()

if __name__ == '__main__':
    asyncio.run(main())
```

Author: hmm so steps 1 and 2 should be combined? In which case tablestep formatting probably doesn't make sense for this whole section anymore eh?

Author: This feels a bit unnecessarily hand-holdy to me since they need to customize the script for their use case anyway, with different tags, part ID, filepath, and API keys

Reviewer: It means fewer pages they need to open. It is much nicer to copy an entire script and swap out a few things than to have to piece things together from different pages. You can have copying it be the first step and  finding the ids the second

Author: Okay I'm going to leave the order as-is and add this code as expanders. I think it still makes sense that way. But, I'm having a code formatting issue when I put them in expanders https://viaminc.slack.com/archives/C055G9S1W3Y/p1725493005478279

```

{{% /tab %}}
{{% tab name="Upload all files in a directory" %}}

> To upload all the files in a directory, you can use the same [`file_upload_from_path`](/appendix/apis/data-client/#fileuploadfrompath) method inside a `for` loop, for example:

Reviewer: ```suggestion
To upload all the files in a directory, you can use the [`file_upload_from_path`](/appendix/apis/data-client/#fileuploadfrompath) method inside a `for` loop, for example:
```

> ```python {class="line-numbers linkable-line-numbers"}

Reviewer: same here

This is useful if you want to capture images for training machine learning models on the go.

### Prerequisites

{{< expand "Download the Viam mobile app and sign into your Viam account" >}}

Install the mobile app from the [App Store](https://apps.apple.com/vn/app/viam-robotics/id6451424162) or [Google Play](https://play.google.com/store/apps/details?id=com.viam.viammobile&hl=en&gl=US).

<a href="https://apps.apple.com/vn/app/viam-robotics/id6451424162" target="_blank">
  <img src="https://github.com/viamrobotics/docs/assets/90707162/a470b65d-1b97-412f-9f97-daf902f2f053" width="200px" alt="apple store icon" class="center-if-small" >
</a>

<a href="https://play.google.com/store/apps/details?id=com.viam.viammobile&hl=en&gl=US" target="_blank">
  <img src="https://github.com/viamrobotics/docs/assets/90707162/6ebd6960-08c5-41d4-81f9-42293fbfdfd4" width="200px" alt="google play store icon" class="center-if-small" >
</a>

{{< /expand >}}

### Instructions

{{< table >}}
{{% tablestep link="/services/data/" %}}
**1. Navigate to your machine**

> Select an organization by clicking on the menu icon in the top left corner and tapping an organization.

Reviewer: ```suggestion
In the Viam app, select an organization by clicking on the menu icon in the top left corner and tapping an organization.
```

Author: ```suggestion
In the Viam mobile app, log in, then select an organization by clicking on the menu icon in the top left corner and tapping an organization.
```

Tap the **Locations** tab and select a location, then select the machine you want your data to be associated with.

{{% /tablestep %}}
{{% tablestep %}}
**2. Upload images**

Tap the menu button marked "**...**" in the upper right corner.
Tap **Upload Images**.

Select each image you want to upload, then tap **Add**.

The uploaded images metadata will contain the machine part you selected.
However, the uploaded images will not be associated with a component or method.

{{% /tablestep %}}

> {{< /table >}}

Reviewer: Add a next steps section please. I think these may be useful:
- https://docs-test.viam.dev/3367/how-tos/image-data/
- https://docs-test.viam.dev/3367/how-tos/deploy-ml/
